### PR TITLE
Refactor Sign Languages and Language Markers

### PR DESCRIPTION
**Pull request by Mnemotechnician**
# Description
This refactors #510 and #553. #553 specifically was reverted and re-implemented from scratch. As a consequence to all of this, the chat system was refactored a bit too, hopefully for the best.

Changes:
- InGameICChatType, InGameOOCChatType, ChatTransmitRange were all moved to shared and made serializable
- Added a method to wrap whisper messages to reduce code duplication in chat system
- Both WrapPublicMethod and WrapWhisperMessage call the same generic WrapMessage method, which allows to add speech verbs to whispers and more. That method is also fully responsible for adding language markers and deducing speech verbs now.
- Everything related to speech was moved out of LanguagePrototype and into SpeechOverrideInfo. LanguagePrototype now holds an instance of that.
- Added AllowRadio, RequireSpeech, ChatTypeOverride, SpeechVerbOverrides, MessageWrapOverrides to SpeechOverrideInfo, all of which are used in implementing the sign language.
- Suffered a lot

# TODO
- [X] Cry
- [X] Fix the sign language not displaying properly over the character.
- [X] Find a way to circumvent being unable to speak??

<details><summary><h1>Media</h1></summary><p>

image

See below

</p></details>

# Changelog
No cl no fun

### CODE REVIEW

File: Content.Shared/Chat/SharedChatSystem.cs [hunk 1]
```diff
@@ -243,3 +244,42 @@ public static string GetStringInsideTag(ChatMessage message, string tag)
...
+///     InGame IC chat is for chat that is specifically ingame (not lobby) but is also in character, i.e. speaking.
+/// </summary>
+// ReSharper disable once InconsistentNaming
+[Serializable, NetSerializable]
+public enum InGameICChatType : byte
+{
+    Speak,
+    Emote,
+    Whisper,
+    Telepathic //Nyano - Summary: adds telepathic as a type of message users can receive.
```
Comment (DangerRevolution) [inline Content.Shared/Chat/SharedChatSystem.cs:258]:
```suggestion
    Telepathic
```

**Comment by Mnemotechnician** [comment]
Okay, I fixes and it works properly now
image

image

image

image

**Comment by FoxxoTrystan** [comment]
Hows does it looks like when the other does not know "Sign Language"?

**Comment by FoxxoTrystan** [comment]
Do you think we could also provide an override for the bubble type, its should be possible as i have seen but would it be easly to imprement a bubble type.

**Comment by Mnemotechnician** [comment]
This should now be ready for review, this is how it looks in-game:


https://github.com/user-attachments/assets/c36ee9b0-f767-49c2-80f9-1a598d6c5815

File: Resources/Prototypes/Language/languages.yml [hunk 2]
```diff
@@ -576,3 +575,22 @@
...
+      Speak: chat-sign-language-message-wrap
+      Whisper: chat-sign-language-whisper-wrap
+    speechVerbOverrides:
+    - chat-speech-verb-sign-1
+    - chat-speech-verb-sign-2
+    - chat-speech-verb-sign-3
+  obfuscation:
+    !type:ReplacementObfuscation
+    replacement:
+    - something
```
Comment (DangerRevolution) [inline Resources/Prototypes/Language/languages.yml:596]:
```suggestion
    - something
    - a cryptic message to you
    - a signal to you
    - a message
    - # Blank - Does this work?
    - a rude expression to you
    - a sad expression to you
    - a happy expression to you
```

**Review by DangerRevolution — DISMISSED** [review]
needs moar; in a future PR would you be able to make sign language a trait?

File: Resources/Prototypes/Language/languages.yml (hunk 2 above)
Comment (DEATHB4DEFEAT) [reply Resources/Prototypes/Language/languages.yml:596]: The blank won't work, there are probably other ways to do it though

**Comment by Mnemotechnician** [comment]
> needs moar; in a future PR would you be able to make sign language a trait?

Yes but that's not within the scope of this PR. I think mimes should be able to speak it naturally, too.

File: Resources/Prototypes/Language/languages.yml (hunk 2 above)
Comment (FoxxoTrystan) [reply Resources/Prototypes/Language/languages.yml:596]: Please just leave it simple, we dont need all of those.

**Review by VMSolidus — APPROVED** [review]

**Comment by FoxxoTrystan** [comment]
> needs moar; in a future PR would you be able to make sign language a trait?

Again we want to do a Language Menu but temp we got do a sign language trait **TEMP**

**Comment by VMSolidus** [comment]
> > needs moar; in a future PR would you be able to make sign language a trait?
> 
> Again we want to do a Language Menu but temp we got do a sign language trait **TEMP**

I do agree that it absolutely is essential for sign language that for it to really be "Playable", we absolutely must have a trait for people other than Mute & Mime to buy it(Actually I don't like the idea that Mimes get it for free, they absolutely shouldn't. But that's because I hate mimes, think they're offensive, and think that they should suffer because they're people who voluntarily choose to be mute.). Obviously, the trait could just be trivially removed once an actual language menu is coded. 

My biggest issue right now with tying it to the Mute trait. Is that the mute trait is absolutely dogshit, to such an extreme degree that most people who play "Mute" characters do so without taking the trait, myself included.

**Comment by VMSolidus** [comment]
Other gripes. Sign Language "Speech" should have quotation marks, to indicate that it's actually talking and not just a funky capitalized special emote text.

**Review by OldDanceJacket — APPROVED** [review]

File: Content.Server/Chat/Systems/ChatSystem.cs [hunk 3]
```diff
@@ -527,33 +520,26 @@ private void OnGameChange(GameRunLevelChangedEvent ev)
             {
                 // Scenario 1: the listener can clearly understand the message
                 result = perceivedMessage;
-                wrappedMessage = Loc.GetString("chat-manager-entity-whisper-wrap-message",
-                    ("color", language.Color ?? Color.Gray),
-                    ("entityName", name),
-                    ("message", FormattedMessage.EscapeText(result)));
+                wrappedMessage = WrapWhisperMessage(source, "chat-manager-entity-whisper-wrap-message", name, perceivedMessage, language);
```
Comment (FoxxoTrystan) [inline Content.Server/Chat/Systems/ChatSystem.cs:523]: wrappedMessage = WrapWhisperMessage(source, "chat-manager-entity-whisper-wrap-message", name, perceivedMessage, language);

we should send result not perceivedMessage

**Pull request merged**